### PR TITLE
Build up token authentication middlewares

### PR DESCRIPTION
**Pull request by YongyanDeng**
Build up JWT authentication middleware.

### CODE REVIEW

File: server/middleware/auth.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      } catch (err) {
+            return next(err);
+      }
+}
+
+exports.vendorVerify = async function (req, res, next) {
+      try {
+            const token = req.headers.authorization.split(' ')[1];
+            const decoded = jwt.verify(token, process.env.JWT_SECRET_KEY);
+            if (decoded && decoded.category === 'VENDOR') return next();
```
Comment (JieqLuo) [inline server/middleware/auth.js:35]: where is the decoded.category 'VENDOR' defined

File: server/index.js [hunk 2]
```diff
@@ -19,6 +21,16 @@ app.use((req, res, next) => {
       next(err);
 });
 
+app.get('/api/products', loginVerify, vendorVerify, async function (req, res, next) {
```
Comment (JieqLuo) [inline server/index.js:24]: is this the authentication process for only vendor? do we use it for user authorization?

File: server/middleware/auth.js (hunk 1 above)
Comment (YongyanDeng) [reply server/middleware/auth.js:35]: category 'VENDOR' is defined in ``/models/user.js``. In the pre-hook:
```
// vendor category
userSchema.pre('save', async function (next) {
      try {
            if (this.email.includes('@vendor.com')) this.category = 'VENDOR';
            return next();
      } catch (err) {
            return next(err)
      }
});
```
And I wrap it into the token with ``jwt.sign`` in ``/handlers/auth.js``. The token will be saved in the ``localStorage`` of the React and then will be used in ``decoded`` to check if this account is vendor or user.

File: server/index.js (hunk 2 above)
Comment (YongyanDeng) [reply server/index.js:24]: There are 3 auth middlewares in ``/middleware/auth.js``. ``loginVerify`` is for all accounts to check if they have signed in or not. ``userVerify`` is for USER identity, and ``vendorVerify`` is for VENDOR identity.

**Review by JieqLuo — APPROVED** [review]

**Pull request merged**